refactor/profile

mohammed-akkad opened the pull request:
This pull request introduces a comprehensive refactor to the application's settings and localization infrastructure. The main goals are to modernize language and theme management by leveraging Jetpack DataStore (replacing SharedPreferences), centralize locale handling using Android's new APIs, and make theme selection reactive through Kotlin Flows. Additionally, the codebase is updated to use `AppCompatActivity` and Jetpack Compose idioms throughout.

The most important changes include:

**Localization and Language Management:**

- Added `locales_config.xml` and updated `AndroidManifest.xml` to declare supported locales and integrate Android's locale management, including the `AppLocalesMetadataHolderService` for automatic locale storage. 

- Introduced the `InstallSavedAppLanguage` composable to apply the saved app language at startup using the latest Android APIs and Compose patterns, replacing manual context wrapping. 

- Removed old locale management code from `MainActivity` and `AuthActivity`, simplifying activity initialization.

**Settings Persistence and Data Layer:**

- Migrated settings storage from `SharedPreferences` to Jetpack DataStore, updating `SettingLocalDataSourceImp` and related DI modules to support reactive, type-safe preference management.

- Extended the settings data model and repository interfaces to support onboarding completion, theme, and restriction management as suspend functions and Flows.

**Theme Management:**

- Updated the `AflamiTheme` composable to accept a `Flow<Boolean>` for dark theme selection, enabling dynamic theme changes based on persisted user preferences. Adjusted previews and usages accordingly. 

- Changed the app's base theme to use `Theme.AppCompat.DayNight.NoActionBar` for better compatibility with dynamic theming. 

**Architecture and Code Quality:**

- Refactored `MainActivity` and `AuthActivity` to use `AppCompatActivity` and Compose idioms, removing legacy code and improving maintainability.

These changes modernize the app's settings and localization approach, improve user experience with seamless language and theme changes, and align the codebase with current Android development standards.

[Screencast from 08-11-2025 01:58:42 PM.webm](https://github.com/user-attachments/assets/91264b25-ad56-42c3-9624-3e804e1169a4)
[Screencast from 08-11-2025 02:49:02 PM.webm](https://github.com/user-attachments/assets/c2f05b99-ad01-4db9-9493-579c08470f26)

mohammed-akkad committed:
update aflami theme

mohammed-akkad committed:
update data source module

mohammed-akkad committed:
update game card

mohammed-akkad committed:
update strings

mohammed-akkad committed:
update setting use case

mohammed-akkad committed:
update setting repository

mohammed-akkad committed:
update selection card

mohammed-akkad committed:
update app slider

mohammed-akkad committed:
update continue watching

mohammed-akkad committed:
update media

mohammed-akkad committed:
add restriction dialog

mohammed-akkad committed:
update search

mohammed-akkad committed:
update home screen

mohammed-akkad committed:
update list details screen

mohammed-akkad committed:
update watch history screen

mohammed-akkad committed:
update my rating screen

mohammed-akkad committed:
use data store instead of shared preferences

mohammed-akkad committed:
update safe image viewer

mohammed-akkad committed:
refactor language

mohammed-akkad committed:
refactor activity

mohammed-akkad committed:
Merge remote-tracking branch 'origin/develop' into refactor/profile

# Conflicts:
#	app/src/main/java/com/paris_2/aflami/MainActivity.kt
#	app/src/main/res/values/themes.xml

mohammed-akkad committed:
update release info

mohammed-akkad committed:
delete log

mohammed-akkad committed:
refactor main activity

mohammed-akkad committed:
refactor setting repository test

mohammed-akkad committed:
update setting local data source test

Joseph-Sameh-0 commented on feature/home/homeUi/src/main/java/com/feature/home/homeUi/screen/topRatingMovies/TopRatingMoviesViewModel.kt:
```diff
@@ -1,28 +1,67 @@
     }
 
+    private fun getRestriction() {
+        viewModelScope.launch {
+            val restriction = settingsUseCase.getRestriction()
+            emitState(
+                screenState.value.copy(
+                    contentRestriction = ContentRestriction.valueOf(restriction)
+                )
+            )
+            when (screenState.value.contentRestriction) {
+                ContentRestriction.STRICT -> emitState(
+                    screenState.value.copy(
+                        nsfwThreshold = 0.8f,
+                        genderThreshold = 0.6f
+                    )
```
why you don't use try to execute?

mohammed-akkad replied:
done

Joseph-Sameh-0 commented on feature/lists/listsUi/src/main/java/com/feature/lists/listsUi/screens/listDetails/ListDetailsViewModel.kt:
```diff
@@ -21,16 +23,51 @@ class ListDetailsViewModel @Inject constructor(
+                )
+            )
+            when (screenState.value.contentRestriction) {
+                ContentRestriction.STRICT -> emitState(
```
same here 
if you used try to execute you don'tneed to make viewModelScope

Joseph-Sameh-0 commented on feature/mediaDetails/mediaDetailsUi/src/main/java/com/feature/mediaDetails/mediaDetailsUi/ui/screen/main/MainViewModel.kt:
```diff
@@ -0,0 +1,17 @@
+class MainViewModel @Inject constructor(
+    private val settingsUseCase: SettingsUseCase,
+) : ViewModel(){
+
+
+    fun getLastSelectedAppLanguage() =
+        settingsUseCase.getLanguage()
+
+}
```
why all this space? 😂

Joseph-Sameh-0 commented on feature/mediaDetails/mediaDetailsUi/src/main/java/com/feature/mediaDetails/mediaDetailsUi/ui/screen/movie/details/MovieDetailsViewModel.kt:
```diff
@@ -98,11 +101,45 @@ class MovieDetailsViewModel @Inject constructor(
+                screenState.value.copy(
+                    contentRestriction = ContentRestriction.valueOf(restriction)
+                )
+            )
```
same for try to execute

Joseph-Sameh-0 commented on feature/mediaDetails/mediaDetailsUi/src/main/java/com/feature/mediaDetails/mediaDetailsUi/ui/screen/tvShow/details/TvShowDetailsViewModel.kt:
```diff
@@ -98,10 +101,44 @@ class TvShowDetailsViewModel @Inject constructor(
+            val restriction = settingsUseCase.getRestriction()
+            updateState(
+                screenState.value.copy(
+                    contentRestriction = ContentRestriction.valueOf(restriction)
```
same for try to execute

Joseph-Sameh-0 commented on feature/onboarding/onboardingUi/src/main/java/com/feature/onboarding/onboardingUi/ui/main/MainViewModel.kt:
```diff
@@ -0,0 +1,17 @@
+class MainViewModel @Inject constructor(
+    private val settingsUseCase: SettingsUseCase,
+) : ViewModel(){
+
+
+    fun getLastSelectedAppLanguage() =
+        settingsUseCase.getLanguage()
+
```
format the code  😂

Joseph-Sameh-0 commented on feature/profile/profileUi/src/main/java/com/feature/profile/profileUi/screen/main/InstallSavedAppLanguage.kt:
```diff
@@ -0,0 +1,41 @@
+){
+
+    LaunchedEffect(key1 = Unit){
+        mainViewModel.getLastSelectedAppLanguage().collectLatest {language ->
```
```suggestion
        mainViewModel.getLastSelectedAppLanguage().collectLatest { language ->
```
format the code 😂

Joseph-Sameh-0 commented on feature/profile/profileUi/src/main/java/com/feature/profile/profileUi/screen/watchHistory/WatchHistoryUiState.kt:
```diff
@@ -6,8 +6,18 @@ data class WatchHistoryUiState(
+    val genderThreshold: Float = 0.6f
 )
 
+enum class ContentRestriction() {
+    STRICT,
+    MODERATE,
+    OFF
+}
```
```suggestion
enum class ContentRestriction() {
    Strict,
    Moderate,
    Off
}
```

it's recommended

Joseph-Sameh-0 commented on feature/profile/profileUi/src/main/java/com/feature/profile/profileUi/screen/ProfileViewModel.kt:
```diff
@@ -38,6 +40,20 @@ class ProfileViewModel @Inject constructor(
     }
 
+    private fun getRestriction() {
+        viewModelScope.launch {
+            val restriction = settingsUseCase.getRestriction()
+            updateState(
+                screenState.value.copy(
+                    profile = screenState.value.profile.copy(
+                        contentRestriction = ContentRestriction.valueOf(restriction)
+                    )
+                )
+            )
+        }
+
+    }
```
try to execute

Joseph-Sameh-0 commented on feature/profile/profileUi/src/main/java/com/feature/profile/profileUi/screen/ProfileViewModel.kt:
```diff
@@ -68,14 +84,20 @@ class ProfileViewModel @Inject constructor(
-        )
+            }
+
+        }
```
try to execute

Joseph-Sameh-0 commented on feature/profile/profileUi/src/main/java/com/feature/profile/profileUi/screen/ProfileViewModel.kt:
```diff
@@ -111,7 +139,10 @@ class ProfileViewModel @Inject constructor(
-        settingsUseCase.setTheme(appearance == Appearance.DARK)
+        viewModelScope.launch {
+            settingsUseCase.setTheme(appearance == Appearance.DARK)
+        }
```
try to execute

Joseph-Sameh-0 commented on feature/profile/profileUi/src/main/java/com/feature/profile/profileUi/screen/ProfileViewModel.kt:
```diff
@@ -201,6 +232,12 @@ class ProfileViewModel @Inject constructor(
+    override fun onRestrictionSelected(contentRestriction: ContentRestriction) {
+            viewModelScope.launch {
+                settingsUseCase.setRestriction(contentRestriction.name)
+            }
```
try to execute

mohammed-akkad committed:
use try to execute in view model

mohammed-akkad committed:
refactor main view model

mohammed-akkad committed:
refactor install saved app language

mohammed-akkad committed:
refactor view model

mohammed-akkad committed:
update profile view model

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/profile

mohammed-akkad committed:
Update ContinueWatchingViewModelTest.kt

mohammed-akkad committed:
Update HomeScreenViewModelTest.kt

mohammed-akkad committed:
Update HomeScreenViewModelTest.kt

mohammed-akkad committed:
Update ContinueWatchingViewModelTest.kt

mohammed-akkad committed:
Update TopRatingMoviesViewModelTest.kt

Joseph-Sameh-0 committed:
Refactor: Update ContinueWatchingViewModelTest to include content restriction handling

Joseph-Sameh-0 committed:
Refactor: Enhance HomeScreenViewModelTest with consistent restriction handling

Joseph-Sameh-0 committed:
Refactor: Integrate content restriction handling in TopRatingMoviesViewModelTest

Joseph-Sameh-0 committed:
Refactor: Inject SettingsUseCase into ViewModel tests

Updated TvShowDetailsViewModelTest, MovieDetailsViewModelTest, and WatchHistoryViewModelTest to include SettingsUseCase as a constructor dependency. This change ensures that the ViewModels are instantiated correctly with all required dependencies for testing.

Joseph-Sameh-0 committed:
Refactor: Inject SettingsUseCase into MyRatingViewModelTest

mohammed-akkad committed:
Update FindByActorViewModelTest.kt

mohammed-akkad committed:
Update SearchViewModelTest.kt

mohammed-akkad committed:
Update WorldTourViewModelTest.kt

mohammed-akkad committed:
Merge branch 'develop' into refactor/profile

Joseph-Sameh-0 committed:
Refactor: Correctly inject settingsUseCase in WorldTourViewModelTest

Joseph-Sameh-0 committed:
Refactor: Update SearchViewModelTest, WorldTourViewModelTest, and FindByActorViewModelTest

- Updated SearchViewModelTest to relax mocks for use cases and fix argument order in onApplyFilterButtonClick.
- Refactored WorldTourViewModelTest to use relaxed mocks, clear mocks in tearDown, and ensure proper test isolation for onMediaCardClick.
- Updated FindByActorViewModelTest to use relaxed mocks for SettingsUseCase and ensure proper initialization in onMediaCardClick test.

Joseph-Sameh-0 committed:
Refactor: Mock settingsUseCase restriction in SearchViewModelTest

Joseph-Sameh-0 committed:
Update WorldTourViewModelTest.kt

Joseph-Sameh-0 committed:
Refactor: Improve error handling and data processing in SearchViewModel and PagingSource

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/profile

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/profile

Joseph-Sameh-0 committed:
Revert "Refactor: Improve error handling and data processing in SearchViewModel and PagingSource"

mohammed-akkad committed:
Merge remote-tracking branch 'origin/develop' into refactor/profile

# Conflicts:
#	app/release-info.txt
#	app/src/main/java/com/paris_2/aflami/di/DataSourceModule.kt
#	datasource/local/user/src/main/java/com/paris_2/dataSource/local/user/SettingLocalDataSourceImpl.kt
#	datasource/local/user/src/test/java/com/paris_2/dataSource/local/user/SettingLocalDataSourceImplTest.kt
#	repository/user/src/main/java/com/paris_2/repository/user/repository/SettingRepositoryImpl.kt

mohammed-akkad committed:
update data soruce module

mohammed-akkad committed:
Merge remote-tracking branch 'origin/develop' into refactor/profile

# Conflicts:
#	feature/home/homeUi/src/main/java/com/feature/home/homeUi/screen/home/HomeScreenUIState.kt
#	feature/mediaDetails/mediaDetailsUi/src/main/res/values-ar/strings.xml
#	feature/mediaDetails/mediaDetailsUi/src/main/res/values/strings.xml

Joseph-Sameh-0 commented on safeimageviewer/src/main/java/com/designSystem/safeimageviewer/SafeImageViewer.kt:
```diff
@@ -31,10 +31,10 @@ fun SafeImageViewer(
     nsfwThreshold: Float = 0.8f,
     genderThreshold: Float = 0.6f,
+    blurNSFW: Boolean = nsfwThreshold != 0f,
+    blurFemales: Boolean = nsfwThreshold != 0f,
```
```suggestion
    blurFemales: Boolean = genderThreshold  != 0f,
```

ahmedNaser7 commented on feature/authentication/authenticationUi/src/main/java/com/feature/authentication/authenticationUi/screen/main/InstallSavedAppLanguage.kt:
```diff
@@ -0,0 +1,38 @@
+import kotlinx.coroutines.withContext
+
+@Composable
+fun InstallSavedAppLanguage(
```
why composable function take a context ?

mohammed-akkad replied:
becuse need it when change language

ahmedNaser7 commented on feature/home/homeUi/src/main/java/com/feature/home/homeUi/common/ContentRestriction.kt:
```diff
@@ -0,0 +1,7 @@
+package com.feature.home.homeUi.common
+
+enum class ContentRestriction() {
```
content restriction could be more easier if you make the safe image viewr deal with settting local data

mohammed-akkad replied:
mmmm, i dont know if it the best practiceto call local data source in save image viewer, what about you?

Joseph-Sameh-0 replied:
the safe Image viewer like a library to use 
it shouldn't know about our data source or the application at all

mohammed-akkad committed:
Update SafeImageViewer.kt

Joseph-Sameh-0 committed:
Merge branch 'develop' into refactor/profile
